composer/utils/module_surgery.py

```diff
@@ -71,41 +65,62 @@ def replace_module_classes(
 ) -> Dict[torch.nn.Module, torch.nn.Module]:
     """Modify model in-place by recursively applying replacement policies.
 
-    Examples:
-        The following policy::
-
-            policies = {
-                nn.Conv2d: lambda x, idx: nn.Linear(16, 32),
-                nn.MaxPool2d: lambda x, idx: nn.AvgPool2d(3, stride=2),
-                nn.Linear: lambda x, idx: nn.Linear(16, 64) if x.in_features == 32 else None
-            }
+    .. rubric:: Example
+
+    The following example replaces all convolution layers with linear layers, and
+    linear layers will be optionally replaced depending on the number of input features.
+
+    .. testsetup::
+    
+        from composer.utils.module_surgery import replace_module_classes
+    
+    .. doctest::
+
+        >>> from torch import nn
+        >>> module = nn.Sequential(
+        ... nn.Conv2d(1, 32, 3, 1),
+        ... nn.ReLU(),
+        ... nn.MaxPool2d(2),
+        ... nn.Flatten(),
+        ... nn.Linear(5408, 128),
+        ... nn.ReLU(),
+        ... nn.LogSoftmax(dim=1),
+        )
+        >>> policies = {
+        ...    nn.Conv2d: lambda x, idx: nn.Linear(16, 32),
+        ...    nn.Linear: lambda x, idx: nn.Linear(16, 64) if x.in_features == 32 else None
+        ... }
+        >>> replace_module_classes(module, policies)
+        {Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1)): Linear(in_features=16, out_features=32, bias=True), ...}
```

Comment:
What happens if you have a replacement policy such that the new/replacement module in one policy is the old/module to be replaced in another policy? For example
```
policies = {
    nn.Conv2d: lambda x, idx: nn.Linear(),
    nn.Linear: lambda x, idx: nn.LayerNorm()
}
```
Would all the `Conv2d` layers be `LayerNorm` layers? I assume so. We should probably note this explicitly in the documentation. Something like "replacements are performed sequentially, so the order of replacement policies matters" and then provide an example like the above where the desired behavior (`Conv2d` -> `Linear`) isn't happening and `Conv2d` -> `LayerNorm`) is happening instead.

I suppose if ordering is important than the policy should also be an OrderedDict?

Reply:
Good points....I am updating the example to show recursion, and yes, policies should be an `OrderedDict`, since policies are applied in order for each module.

Reply:
Well, technically, each module is a unique type so it should only match on one key. But, with inheritance, it is possible for a module to match on multiple keys, in which case yes ordering is important.